tests/php/Forms/FormFieldTest.php

```diff
@@ -455,6 +474,99 @@ public function testGetSchemaStateWithFormValidation()
+        $form->validationResult();
+        $schema = $field->getSchemaState();
+        $this->assertEquals(
+            'A test error message',
+            $schema['message']['value']
+        );
+    }
+
+    public function testValidationExtensionHooksAreCalledOnFormFieldSubclasses()
+    {
+        // Can't use a provider for this due to class manifest weirdness
+        $formFieldClasses = ClassInfo::subclassesFor(FormField::class, false);
+        foreach ($formFieldClasses as $formFieldClass) {
+            $reflection = new ReflectionClass($formFieldClass);
+            // Skip abstract classes like MultiSelectField
+            if ($reflection->isAbstract()) {
+                continue;
+            }
+
+            switch ($formFieldClass) {
+                case NullableField::class:
+                case CompositeField::class:
+                case FieldGroup::class:
+                case PopoverField::class:
+                    $args = [TextField::create('Test2')];
+                    break;
+                case SelectionGroup_Item::class:
+                    $args = ['Test', [TextField::create('Test2')]];
+                    break;
+                case ToggleCompositeField::class:
+                    $args = ['Test', 'Test', TextField::create('Test2')];
+                    break;
+                case PrintableTransformation_TabSet::class:
+                    $args = [Tab::create('TestTab', 'Testtab', TextField::create('Test2'))];
+                    break;
+                case TreeDropdownField::class:
+                case TreeDropdownField_Readonly::class:
+                    $args = ['Test', 'Test', Group::class];
+                    break;
+                case PermissionCheckboxSetField::class:
+                case PermissionCheckboxSetField_Readonly::class:
+                    $args = ['Test', 'Test', Permission::class, 'Test'];
+                    break;
+                case SelectionGroup::class:
+                    $args = ['Test', []];
+                    break;
+                case GridField_FormAction::class:
+                    $args = [GridField::create('GF'), 'Test', 'Test label', 'Test action name', []];
+                    break;
+                case GridState::class:
+                    $args = [GridField::create('GF')];
+                    break;
+                default:
+                    $args = ['Test', 'Test label'];
+            }
+
+            echo $formFieldClass . "\n";
+            ob_flush();
+
+            $mock = $this->getMockBuilder($formFieldClass)
+                ->setConstructorArgs($args)
+                ->setMethods(['extendValidationResult'])
+                ->getMock();
+
+            $mock->expects($this->once())
+                ->method('extendValidationResult')
+                ->will($this->returnValue(true));
+
+            $this->assertTrue($mock->validate(new RequiredFields()));
```

Comment:
```suggestion
            try {
                $isValid = $mock->validate(new RequiredFields());
            } catch (ExpectationFailedException $e) {
                $this->fail("Expectation failed for '$formFieldClass' class: {$e->getMessage()}");
            }
            $this->assertTrue($isValid, "$formFieldClass should be valid");
```
This should make it clear which field was being tested when the `$mock->expects($this->once())` expectation fails - and added message for the `assertTrue()` as well for good measure.

Reply:
Obviously also need to add `use PHPUnit\Framework\ExpectationFailedException;`

Reply:
This doesn’t work 😢 looks like mock expectations aren’t checked until after the entire test has run

Reply:
Oh, weird. It worked for me locally when I changed `$this->once()` to `$this->never()`...

Reply:
That’s bizarre - it does work for `->never()`, but not for `once()` (I’m testing making the `once()` expectatuin fail by manually removing the `->extendValidationResult()` call from `EmailField`)...

Reply:
Oh, weird.... We might just not be able to add the message for that expectation then. 

This might be worth trying: https://stackoverflow.com/a/56677638 explicitly invoking an assertion for it, if that's still valid API.

But not the end of the world if we can't get a nice message in there. At least if it fails again in the future we can add a temporary debug line in like you did this time to see what class failed it.

Reply:
Looks like that API has been removed. Calling this method (`$mock->__phpunit_verify())` works, but it’s not officially supported API so it’s risky:

https://github.com/sebastianbergmann/phpunit/blob/ddf1b957bd90342cecdaec56f62e98d410187cd9/src/Framework/MockObject/Api/Api.php#L72-L79

Ideally we’d do this with a dataProvider, as that does seem to include the class name in the error, but dataProviders are fetched _so_ early that our manifests aren’t ready, and even `get_declared_classes()` won’t work without manually querying autoloaders and including files... nasty.

Reply:
Oh wait, I’ve found something else that’ll work:

```php
$mock = $this->getMockBuilder($formFieldClass)
    ->setConstructorArgs($args)
    ->onlyMethods(['extendValidationResult'])
    ->getMock();

$mock->expects($invocationRule = $this->once())
    ->method('extendValidationResult')
    ->will($this->returnValue(true));

$isValid = $mock->validate(new RequiredFields());
$this->assertTrue($isValid, "$formFieldClass should be valid");

try {
    $invocationRule->verify();
} catch (\Exception $e) {
    $this->fail("Expectation failed for '$formFieldClass' class: {$e->getMessage()}");
}
```